File: Codes/1807-evaluate-the-bracket-pairs-of-a-string.py
'''
Author: LetMeFly
Date: 2023-01-12 22:23:47
LastEditors: LetMeFly
LastEditTime: 2023-01-12 22:40:28
'''
from typing import List

if False:
    class Solution:
        def evaluate(self, s: str, knowledge: List[List[str]]) -> str:
            ma = {}
            for v in knowledge:
                ma[v[0]] = v[1]
            ans = ""
            for i in range(0, len(s)):  # 不可，因为下面对i的更改无法作用到这里
                if (s[i] == '('):
                    to = i + 1
                    while s[to] != ')':
                        to += 1
                    ans += ma.get(s[i + 1 : to])
                    i = to
                else:
                    ans += s[i]
            return ans

# True
class Solution:
    def evaluate(self, s: str, knowledge: List[List[str]]) -> str:
        ma = {}
        for v in knowledge:
            ma[v[0]] = v[1]
        ans = ""
        i = 0
        while i < len(s):
            if (s[i] == '('):
                to = i + 1
                while s[to] != ')':
                    to += 1
                ans += ma.get(s[i + 1 : to], "?")
                i = to
            else:
                ans += s[i]
            i += 1
        return ans

File: Codes/1807-evaluate-the-bracket-pairs-of-a-string.cpp
/*
 * @Author: LetMeFly
 * @Date: 2023-01-12 22:15:05
 * @LastEditors: LetMeFly
 * @LastEditTime: 2023-01-12 22:21:28
 */
#ifdef _WIN32
#include "_[1,2]toVector.h"
#endif

class Solution {
public:
    string evaluate(string s, vector<vector<string>>& knowledge) {
        string ans;
        unordered_map<string, string> ma;
        for (auto& v : knowledge) {
            ma[v[0]] = v[1];
        }
        for (int i = 0; i < s.size(); i++) {
            if (s[i] == '(') {
                int to = i + 1;
                while (s[to] != ')') {
                    to++;
                }
                string key = s.substr(i + 1, to - i - 1);
                ans += ma.count(key) ? ma[key] : "?";
                i = to;  // 循环结束后会有i++
            }
            else {
                ans += s[i];
            }
        }
        return ans;
    }
};